scripts/linters/ts_ignore_exceptions.json

```diff
@@ -0,0 +1,145 @@
+{
+  "core/templates/domain/improvements/NeedsGuidingResponsesTaskObjectFactorySpec.ts": [
+    "task_type: '???',"
+  ],
+  "extensions/interactions/LogicProof/static/js/tools/logic-demo-test.controller.spec.ts": [
+    "properties than entries according with lint settings.",
+    "properties than {} according with lint settings."
+  ],
+  "extensions/interactions/CodeRepl/directives/code-repl-validation.service.spec.ts": [
```

Comment:
Can you use `@ts-expect-error` instead for test code (see docs online for why this is best practice)? Also add a check to forbid use of `@ts-expect-error` in non-spec files.

Then you can omit such examples from this file.

Reply:
I think `@ts-expect-error` can also be used in non `spec` files. Also, I've added a check for it along with the `ts-ignore` check since it may also be unnecessarily used to suppress errors.

Also FYI -- The places where we used `@ts-ignore` can be divided into the following categories.

I think we can use `@ts-expect-error` here.
1. Where the control shouldn't reach ideally. Example - (https://github.com/oppia/oppia/pull/10088/files#diff-d28bc2a637cdc32be5d1da90d8f067e2R114-R116)
2. Assigning invalid types to test validations in spec files. Example - (https://github.com/oppia/oppia/pull/10088/files#diff-e4cc090fa5747257556fcac5fc4384ddR142-R143)

and `@ts-ignore` here.
1. More properties are expected while mocking an object. Example - (https://github.com/oppia/oppia/pull/10088/files#diff-7034e0434cc3ccf243d9c8b4f8a4da2fR143-R146)

Reply:
I disagree with using `@ts-expect-error` in non-spec files. As a general rule, we should not be expecting errors in code that runs in production. Also I think that it helps with having clear conventions.

I suggest that we use ts-expect-error for the last two cases you mentioned above (mocking system objects in spec files where there are too many properties to feasibly mock, and assigning invalid types for validation testing). Everything else is ts-ignore. How does that sound?

Also, sorry -- I'm having trouble finding the "where the control shouldn't reach ideally" case. The link you provided doesn't seem to work for me, it just shows the general Files Changed page. Could you please elaborate?

Reply:
@seanlip You can refer -> https://github.com/oppia/oppia/blob/develop/core/templates/services/schema-default-value.service.ts#L114. Here the value of `schema.type` is only supposed to be one of `bool`, `unicode`, `html`, `list`, `dict`, `int`, `float`. And if it is none of them we have the last else block. (Which should ideally never happen)

Reply:
> I suggest that we use ts-expect-error for the last two cases you mentioned above (mocking system objects in spec files where there are too many properties to feasibly mock, and assigning invalid types for validation testing). Everything else is ts-ignore. How does that sound?

I think the `ts-expect-error` should be used in the places where we know an error is going to pop up for sure. So, I think the first type should also use the `ts-expect-error`.

Reply:
I see, thanks. That usage seems fine. Is it a general usage? If not then I think we can treat it as a one-off, similar to MathJax.

You might also be able to get rid of it altogether by throwing an Error, especially if it's not supposed to happen in prod. Just double-check the PR that introduced that line of code to make sure there aren't reasons for logging an error rather than throwing one (maybe link that PR here). If there's no good reason for only soft-logging, then you can just replace that stuff with throwing an error and you won't need the ts-ignore.

Reply:
> I think the ts-expect-error should be used in the places where we know an error is going to pop up for sure. So, I think the first type should also use the ts-expect-error.

See the comment just above this one for ideas on what we do with it.

Also, I do get your point, but I think there is value in introducing a clear and simple codebase-wide convention and it would be better not to muddy it where possible with one exception.

Reply:
I've refactored the else block to log the complete `schema` we are receiving. This way we don't have to use `schema`.